subprojects/native/src/main/java/org/gradle/internal/nativeintegration/filesystem/services/NativePlatformBackedFileMetadataAccessor.java

```diff
@@ -39,14 +40,18 @@ public FileMetadata stat(File f) {
         try {
             stat = files.stat(f, false);
         } catch (NativeException e) {
-            return DefaultFileMetadata.missing(AccessType.DIRECT);
+            throw new UncheckedIOException("Could not stat file " + f.getAbsolutePath(), e);
         }
         AccessType accessType = AccessType.viaSymlink(stat.getType() == FileInfo.Type.Symlink);
         if (accessType == AccessType.VIA_SYMLINK) {
             try {
                 stat = files.stat(f, true);
             } catch (NativeException e) {
-                return DefaultFileMetadata.missing(AccessType.VIA_SYMLINK);
+                // For a symlink cycle, file.exists() returns false when unable to stat the file.
+                if (!f.exists()) {
+                    return DefaultFileMetadata.missing(accessType);
+                }
```

Comment:
So we'll still represent a symlink cycle as a missing file?

Reply:
Yes. That is what happens with all but the native accessors, so I'd do this for the native implementation as well. It is then represented the same way as a broken symlink, which I think is fine.

Reply:
But if you tried to actually open that file, you'd get a `FileSystemLoopException`, no? To me a cycling symlink sounds very much like unreadable content, not a missing file.

Reply:
Good question. When we walk a directory, we currently ignore file system loop exceptions: https://github.com/gradle/gradle/blob/bfcb154efdc7622a51352a37b284a78888857eb8/subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java#L356-L365

Reply:
Alright. I guess your proposal doesn't change behavior, so it's fine if we stay a bit inconsistent for this rare use case.